### ISSUE TITLE
Client crash when reading a Fractionation Unit with an Angular Transducer with Vazkii's ChatFlow installed

### ISSUE THREAD

**Issue by Maxwell-lt** [ReikaKalseki/Reika_Mods_Issues#2049]
Crash report: https://gist.github.com/Maxwell-lt/b1bb28b698a13e8cb877b63df7045eeb

When attempting to measure a Fractionation Unit with the Angular Transducer, the client immediately crashes to desktop. It appears the issue is caused by the `%` symbol in the output. Removing ChatFlow stops the crashes, but the percentage sign fails to print in chat after the efficiency value. No ChatFlow filters were configured, unless there are default filters.

**Comment by Maxwell-lt** [comment]
Without ChatFlow, an error is still thrown, but it doesn't crash the client: https://gist.github.com/Maxwell-lt/c6c88bf97899fca5e0440de42c4e4f93

This implies that the character is not properly escaped.

**Comment by ReikaKalseki** [comment]
[This](https://github.com/ReikaKalseki/RotaryCraft/blob/52595191469098d6be724018f350cb3b3e07f545/Items/Tools/ItemMeter.java#L336) is the unformatted string; I have always used "double %" to display a % sign, and it has always worked in the past...

**Comment by Barhandar** [comment]
Not unique to ChatFlow, going by my log, LiteLoader also crashes it.

**Comment by OvermindDL1** [comment]
Well let's see, source string line from Reika is:
https://github.com/ReikaKalseki/RotaryCraft/blob/52595191469098d6be724018f350cb3b3e07f545/Items/Tools/ItemMeter.java#L336
```java
this.sendMessage(ep, String.format("Efficiency: %.3f%%", te.getYieldRatio()*100));
```
Which will do something like `this.sendMessage(ep, "Efficiency: 25%");` to:
https://github.com/ReikaKalseki/RotaryCraft/blob/52595191469098d6be724018f350cb3b3e07f545/Items/Tools/ItemMeter.java#L371
```java
	private void sendMessage(EntityPlayer ep, String s) {
		//ReikaChatHelper.writeString(s);
		ReikaChatHelper.sendChatToPlayer(ep, s);
	}
```
Which calls into DragonAPI at:
https://github.com/ReikaKalseki/DragonAPI/blob/master/Libraries/IO/ReikaChatHelper.java#L172-L187
```java
	public static void sendChatToPlayer(EntityPlayer ep, String sg) {
		if (sg.length() > 16384) {
			int idx = 0;
			while (idx < sg.length()) {
				String sg2 = sg.substring(idx, Math.min(idx+16384, sg.length()));
				sendChatToPlayer(ep, sg2);
				idx += 16384;
			}
			return;
		}
		String[] parts = sg.split("\\n");
		for (int i = 0; i < parts.length; i++) {
			ChatComponentTranslation chat = new ChatComponentTranslation(parts[i]);
			ep.addChatMessage(chat);
		}
	}
```
First this splits the message if it is too long, which this one is not.
Next it splits on newlines, which since there are none then it only has one result (`parts.length == 1`), which then packs it into a `ChatComponentTranslation` and sends it on to the player via `ep.addChatMessage(chat)`, so looking into `ChatComponentTranslation` shows it is calling the `ChatComponentTranslation(String translationKey, Object... args)` constructor, and this array `args` looks interesting, well let's see what it does internally...
```java
        boolean flag = false;
        Matcher matcher = stringVariablePattern.matcher(format);
        int i = 0;
        int j = 0;

        try
        {
            int l;

            for (; matcher.find(j); j = l)
            {
                int k = matcher.start();
                l = matcher.end();

                if (k > j)
                {
                    ChatComponentText chatcomponenttext = new ChatComponentText(String.format(format.substring(j, k), new Object[0]));
                    chatcomponenttext.getChatStyle().setParentStyle(this.getChatStyle());
                    this.children.add(chatcomponenttext);
                }

                String s2 = matcher.group(2);
                String s = format.substring(k, l);

                if ("%".equals(s2) && "%%".equals(s))
                {
                    ChatComponentText chatcomponenttext2 = new ChatComponentText("%");
                    chatcomponenttext2.getChatStyle().setParentStyle(this.getChatStyle());
                    this.children.add(chatcomponenttext2);
                }
                else
                {
                    if (!"s".equals(s2))
                    {
                        throw new ChatComponentTranslationFormatException(this, "Unsupported format: \'" + s + "\'");
                    }

                    String s1 = matcher.group(1);
                    int i1 = s1 != null ? Integer.parseInt(s1) - 1 : i++;

                    if (i1 < this.formatArgs.length)
                    {
                        this.children.add(this.getFormatArgumentAsComponent(i1));
                    }
                }
            }

            if (j < format.length())
            {
                ChatComponentText chatcomponenttext1 = new ChatComponentText(String.format(format.substring(j), new Object[0]));
                chatcomponenttext1.getChatStyle().setParentStyle(this.getChatStyle());
                this.children.add(chatcomponenttext1);
            }
        }
        catch (IllegalFormatException illegalformatexception)
        {
            throw new ChatComponentTranslationFormatException(this, illegalformatexception);
        }
```
Well it looks like this is poorly reimplementing a worse version of `String.format`, including even the `%` parsing, and we see if `%` is followed by an invalid spec (or no spec) then it throws `throw new ChatComponentTranslationFormatException(this, "Unsupported format: \'" + s + "\'");`, which happens to match the exception that @Maxwell-lt posted above for an otherwise just reikamods-modded game at:
https://gist.github.com/Maxwell-lt/c6c88bf97899fca5e0440de42c4e4f93

Now since Reika's call is passing in, as an example, `"Efficiency: 25%"` then the `%` is not properly escaped.

In other words this is a bug in DragonAPI.  Formatting should be done via the `ChatComponentTranslation` class, so `Reika.DragonAPI.Libraries.IO.ReikaChatHelper.sendChatToPlayer` should be taking an `Object... args` or so as the formatters and those should be passed to this call:
https://github.com/ReikaKalseki/DragonAPI/blob/master/Libraries/IO/ReikaChatHelper.java#L184
```java
ChatComponentTranslation chat = new ChatComponentTranslation(parts[i]);
```
And `String.format` should not be used for translateable messages (since it also harms translateability in addition to having breakages like passing `%` to it) but rather the `ChatComponentTranslation` is what should handle formatting.

**Comment by OvermindDL1** [comment]
> This is the unformatted string; I have always used "double %" to display a % sign, and it has always worked in the past...

@ReikaKalseki So yes, passing a `%%` is the way to do it, however you are calling `String.format` and passing that returned string to 'essentially' another `String.format`, so you need to double encode the `%` to `%%%%`, or just use the `ChatComponentTranslation` class straight (by passing the format args through sendChatToPlayer and so forth) so someone can properly translate the messages instead (which they cannot do to that string as it stands without translating every-single-possible-value passed in to it).

**Comment by chotke** [comment]
Moreover efficiency factor didnt save if reload a game. Its just halved all the time

**Comment by ReikaKalseki** [comment]
> Moreover efficiency factor didnt save if reload a game. Its just halved all the time

Neither relevant to this issue nor unreported.

**Comment by waterlubber** [comment]
Experienced this same crash on DragonRealm with Opis added. [Crash Log](https://hastebin.com/cuhapoziqa.rb)

**Comment by ReikaKalseki** [comment]
Fixed, but this is ugly.
